### PR TITLE
FBVector Formatter

### PR DESCRIPTION
**Pull request by zrthxn**
Currently the vectors look like this

### Integer
```
(folly::fbvector<int, std::allocator<> >) fbvec = size=5 {
  (int) [0] = 1
  (int) [1] = 2
  (int) [2] = 3
  (int) [3] = 4
  (int) [4] = 5
}
```

### Float/Double
```
(folly::fbvector<double, std::allocator<> >) fbvec = size=4 {
  (double) [0] = 1
  (double) [1] = 2
  (double) [2] = 3
  (double) [3] = 4
}
```

### Character
```
(folly::fbvector<char, std::allocator<> >) fbvec = size=3 {
  (char) [0] = 'A'
  (char) [1] = 'B'
  (char) [2] = 'C'
}
```

### Boolean
```
(folly::fbvector<bool, std::allocator<> >) fbvec = size=2 {
  (bool) [0] = true
  (bool) [1] = false
}
```

### CODE REVIEW

File: src/fbvector/fbvector_formatter.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    vec = FBVectorFormatter(valobj)
+    _summary = ""
+    try:
+        if not (valobj.IsValid()):
+            return "<invalid>"
+        
+        size = vec.num_children()
+        _summary += f"size={size}"
+
+        if size > 0:
```
Comment (a20012251) [inline src/fbvector/fbvector_formatter.py:21]: >= 0

File: src/fbvector/fbvector.cpp [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+fbvector<T> FBVector(std::initializer_list<T> arr) {
+  fbvector<T> fbvec(arr);
+  return fbvec;
+}
+
+int main() {
+  FBVector({ 1, 2, 3, 4, 5 });
+  FBVector({ 1.0, 2.0, 3.0, 4.0 });
+  FBVector({ 'A', 'B', 'C' });
+  FBVector({ true, false });
```
Comment (a20012251) [inline src/fbvector/fbvector.cpp:16]: include an empty vector

File: src/fbvector/fbvector_formatter.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            _summary += " {\n"
+            for i in range(size):
+                _summary += f"  {vec.get_child_at_index(i)}\n"
+            _summary += "}"
+            
+        return _summary
+
+    except Exception:
+        # todo: change this
+        return f"size=-1"
```
Comment (a20012251) [inline src/fbvector/fbvector_formatter.py:31]: just return ""

File: src/fbvector/fbvector_formatter.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    def get_child_at_index(self, index):
+        if index < 0:
+            return None
+        if index >= self.num_children():
+            return None
+        try:
+            # offset = index * self.data_size
+            # return self.start.CreateChildAtOffset(
+            #     '[' + str(index) + ']', offset, self.data_type)
+            return lldb.value(self.start)[int(index)].sbvalue
```
Comment (a20012251) [inline src/fbvector/fbvector_formatter.py:97]: why do you have that comment here? which version is used in c++ std's vector implementation?

File: src/fbvector/test_fbvector.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+from os import path
+
+class FollyFBVectorTest(LLDBTestCase):
+    def get_formatter_path(self):
+        curr_dir = path.abspath(path.dirname(path.realpath(__file__)))
+        return path.join(curr_dir, "fbvector_formatter.py")
+
+    def test_fbvector(self):
+        script = """
+b fbvector.cpp:10
```
Comment (a20012251) [inline src/fbvector/test_fbvector.py:11]: don't forget to add tests here

**Review by a20012251 — CHANGES_REQUESTED** [review]
really good start!

File: src/fbvector/test_fbvector.py (hunk 5 above)
Comment (zrthxn) [reply src/fbvector/test_fbvector.py:11]: Yes I will add those

File: src/fbvector/fbvector_formatter.py (hunk 4 above)
Comment (zrthxn) [reply src/fbvector/fbvector_formatter.py:97]: The commented out version is used in the std::vector implementation
The one I've got there is used in the range formatter, and it makes more sense to me.

File: src/fbvector/fbvector.cpp (hunk 2 above)
Comment (zrthxn) [reply src/fbvector/fbvector.cpp:16]: done

File: src/fbvector/fbvector_formatter.py (hunk 4 above)
Comment (a20012251) [reply src/fbvector/fbvector_formatter.py]: Sure, that's fine. Just finish all the tests in python and good to go!

**Comment by zrthxn** [comment]
I have this problem with the `p` command
So with an empty vector, it prints as expected
```
(lldb) frame var fbvec
(folly::fbvector<unsigned int, std::allocator<> >) fbvec = size=0 {}
(lldb) p fbvec
(folly::fbvector<unsigned int, std::allocator<> >) $0 = size=0 {}
```

But with other vectors, the `p` command prints something empty
```
(lldb) frame var fbvec
(folly::fbvector<int, std::allocator<> >) fbvec = size=5 {
  (int) [0] = 1
  (int) [1] = 2
  (int) [2] = 3
  (int) [3] = 4
  (int) [4] = 5
}
(lldb) p fbvec
$1 =
```

**Comment by zrthxn** [comment]
Added tests but these fail because of the above mentioned issue

**Review by a20012251 — APPROVED** [review]
good job! i'll solve on my own the p issue

**Pull request merged**